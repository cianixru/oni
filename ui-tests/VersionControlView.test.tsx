import { shallow } from "enzyme"
import { shallowToJson } from "enzyme-to-json"
import * as React from "react"

import {
    DefaultState,
    VersionControlState,
} from "./../browser/src/Services/VersionControl/VersionControlStore"
import {
    GitStatus,
    SectionTitle,
    VersionControlView,
} from "./../browser/src/Services/VersionControl/VersionControlView"

const noop = () => ({})

jest.mock("./../browser/src/neovim/SharedNeovimInstance", () => ({
    getInstance: () => ({
        bindToMenu: () => ({
            setItems: jest.fn(),
            onCursorMoved: {
                subscribe: jest.fn(),
            },
        }),
    }),
}))

const makePromise = (arg?: any) => Promise.resolve(arg)

jest.mock("../browser/src/UI/components/Sneakable", () => {
    const React = require("react") // tslint:disable-line
    return { Sneakable: () => <div /> }
})

describe("<VersionControlView />", () => {
    const state = { ...DefaultState, activated: true, hasFocus: true }
    const container = shallow(<VersionControlView {...state} getStatus={() => makePromise({})} />)
    it("renders without crashing", () => {
        expect(container.length).toBe(1)
    })

    it("should render an untracked, staged and modified section", () => {
        const sections = container.dive().find(GitStatus).length
        expect(sections).toBe(3)
    })

    it("shouldn't show a section if it has no content", () => {
        const wrapper = shallow(
            <GitStatus
                onClick={noop}
                toggleVisibility={noop}
                visibility={true}
                titleId="modified"
                selectedId="file1"
                icon="M"
                files={null}
            />,
        )
        expect(wrapper.find(SectionTitle).length).toBe(0)
    })

    it("should match the last recorded snapshot unless a change was made", () => {
        const wrapper = shallow(
            <GitStatus
                titleId="modified"
                visibility={true}
                toggleVisibility={noop}
                onClick={noop}
                selectedId="file1"
                icon="M"
                files={["test1", "test2"]}
            />,
        )
        expect(shallowToJson(wrapper)).toMatchSnapshot()
    })

    it("should render the correct number of modified files from the store in the correct section from of the pane", () => {
        const stateCopy = {
            ...DefaultState,
            activated: true,
            hasFocus: true,
            status: {
                currentBranch: null,
                staged: [],
                conflicted: [],
                created: [],
                modified: ["test1", "test2"],
                remoteTrackingBranch: null,
                deleted: [],
                untracked: [],
                ahead: null,
                behind: null,
            },
        }

        const statusComponent = shallow(
            <VersionControlView {...stateCopy} getStatus={() => makePromise({})} />,
        )
            .dive()
            .findWhere(component => component.prop("titleId") === "modified")

        expect(statusComponent.prop("files").length).toBe(2)
    })
})
