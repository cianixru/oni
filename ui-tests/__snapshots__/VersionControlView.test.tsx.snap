// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VersionControlView /> should match the last recorded snapshot unless a change was made 1`] = `
<div>
  <styled.div
    data-test="modified-2"
    isSelected={false}
    onClick={[Function]}
  >
    <Caret
      active={true}
    />
    <styled.h4>
      MODIFIED
    </styled.h4>
    <strong>
      2
    </strong>
  </styled.div>
  <Sneakable
    callback={[Function]}
    key="test1"
  >
    <styled.div
      isSelected={false}
      onClick={[Function]}
    >
      <styled.div>
        <Icon
          name="M"
        />
        <styled.span>
          test1
        </styled.span>
      </styled.div>
    </styled.div>
  </Sneakable>
  <Sneakable
    callback={[Function]}
    key="test2"
  >
    <styled.div
      isSelected={false}
      onClick={[Function]}
    >
      <styled.div>
        <Icon
          name="M"
        />
        <styled.span>
          test2
        </styled.span>
      </styled.div>
    </styled.div>
  </Sneakable>
</div>
`;
